src/util/statistics/counter/mod.rs

```diff
@@ -22,10 +26,12 @@ pub trait Counter {
     fn name(&self) -> &String;
 }
 
+/// A Diffable object could be stateless (e.g. a timer that reads the wall
+/// clock), or stateful (e.g. holds reference to a perf event fd)
 pub trait Diffable {
     type Val;
-    fn current_value() -> Self::Val;
-    fn diff(current: &Self::Val, earlier: &Self::Val) -> u64;
+    fn current_value(&mut self) -> Self::Val;
+    fn diff(&mut self, current: &Self::Val, earlier: &Self::Val) -> u64;
```

Comment:
I think `diff` does not require `self`. But it is okay to include a `self` param here.